File: examples/java/alias0/Forwarder.java
/*
 * $Id: Forwarder.java 5059 2007-11-19 20:01:59Z mquinson $
 *
 * Copyright 2006,2007 Martin Quinson, Malek Cherier         
 * All rights reserved. 
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. 
 */

import simgrid.msg.*;

public class Forwarder extends simgrid.msg.Process {
    
	public void main(String[] args) throws JniException, NativeException 
	{
		Msg.info("hello!");
	
		int aliasCount = args.length;
		Task taskReceived;
		Task finalizeTask;
		BasicTask basicTask;
		
		int taskCount = 0;
		
		while(true) 
		{
			taskReceived = Task.receive();
				
	
			if(taskReceived instanceof FinalizeTask) 
			{
				Msg.info("All tasks have been dispatched. Let's tell everybody the computation is over.");
	
				for (int i = 0; i < aliasCount; i++) 
				{
					finalizeTask = new FinalizeTask();
					finalizeTask.send(args[i]);
				}
				
				break;
			}
			
			basicTask = (BasicTask)taskReceived;
	
			Msg.info("Received \"" + taskReceived.getName() + "\" ");
			
			Msg.info("Sending \"" + taskReceived.getName() + "\" to \"" + args[taskCount % aliasCount] + "\"");
			
			basicTask.send(args[taskCount % aliasCount]);
	
			taskCount++;
		}
	
	
		Msg.info("I'm done. See you!");
	}
}

File: examples/cxx/basic/Forwarder.cxx
#include "Forwarder.hpp"
#include "BasicTask.hpp"
#include "FinalizeTask.hpp"
#include <Host.hpp>
#include <HostNotFoundException.hpp>

#include <Msg.hpp>

MSG_IMPLEMENT_DYNAMIC(Forwarder, Process);


int Forwarder::main(int argc, char** argv)
{
	info("Hello");
	
	int slavesCount = argc;
	
	Host* slaves = new Host[slavesCount];
	
	for (int i = 0; i < argc; i++) 
	{
		try 
		{	      
			slaves[i] = Host::getByName(argv[i]);
		} 
		catch (HostNotFoundException e) 
		{
			error(TEXT_(e.toString()));
			error("Buggy deployment file");
			exit(1);
		}
	}
	
	int taskCount = 0;

	while(true) 
	{
		Task* t = Task::get(0);	
	
		if(t->isInstanceOf("FinalizeTask")) 
		{
			info("All tasks have been dispatched. Let's tell everybody the computation is over.");
	
			for (int cpt = 0; cpt< slavesCount; cpt++) 
				slaves[cpt].put(0, new FinalizeTask());

			delete t;

			break;
		}

		info(TEXT_("Received \"") + TEXT_(t->getName()) + TEXT_("\" "));
	
		info(TEXT_("Sending \"") + TEXT_(t->getName()) + TEXT_("\" to \"") + TEXT_(slaves[taskCount % slavesCount].getName()) + TEXT_("\""));
		
		slaves[taskCount % slavesCount].put(0, t);
	
		taskCount++;
	}
	
	
	info("I'm done. See you!");
	
	delete[] slaves;

	delete this;
      
	return 0;
}